File: src/main/shutdown.ts
// Graceful-shutdown decision logic — kept dependency-free (no Electron) so it
// can be unit-tested (see scripts/test-shutdown.mjs).
//
// Why this matters: shutting a hot pellet grill straight off can let the fire
// smoulder back up the auger toward the hopper — a burnback / hopper fire. The
// safe procedure is to bring the grill down to ~200°F first, THEN power off so
// the controller's fan cool-down can fully extinguish the firepot. This module
// encodes that as a small state machine; main.ts executes the actions and the
// recorder's power-off detection reports completion.

export type ShutdownPhase = 'cooling' | 'finishing' | null;

export interface ShutdownInput {
  moduleIsOn: boolean;
  grillTemp: number | null;
  grillSetTemp: number | null;
  fanState: boolean;
}

export interface ShutdownStep {
  phase: ShutdownPhase;
  action: 'cool' | 'off' | null;              // command main should send
  notice: { title: string; body: string } | null;
}

export interface ShutdownConfig {
  coolAbove: number;   // above this, ramp down before powering off
  coolTarget: number;  // ramp-down setpoint
  coolDoneAt: number;  // once at/below this, proceed to power off
  stallMs: number;
}

export const SHUTDOWN: ShutdownConfig = {
  coolAbove: 250,
  coolTarget: 200,
  coolDoneAt: 210,
  stallMs: 30 * 60_000,
};

// Decide the first step when the user asks to shut down.
export function beginShutdown(inp: ShutdownInput, cfg: ShutdownConfig = SHUTDOWN): ShutdownStep {
  if (inp.moduleIsOn && typeof inp.grillTemp === 'number' && inp.grillTemp > cfg.coolAbove) {
    return {
      phase: 'cooling', action: 'cool',
      notice: {
        title: 'Cooling down before shutdown',
        body: `Bringing the grill from ${inp.grillTemp}° to ${cfg.coolTarget}° first — this prevents a hopper flare-up.`,
      },
    };
  }
  return {
    phase: 'finishing', action: 'off',
    notice: { title: 'Shutting down', body: 'Turning the grill off; the fan will run until the firepot cools.' },
  };
}

// Advance the machine on each fresh grill state.
export function advanceShutdown(phase: ShutdownPhase, inp: ShutdownInput, cfg: ShutdownConfig = SHUTDOWN): ShutdownStep {
  if (phase === 'cooling') {
    if (typeof inp.grillTemp === 'number' && inp.grillTemp <= cfg.coolDoneAt) {
      return {
        phase: 'finishing', action: 'off',
        notice: {
          title: 'Cooled — shutting down',
          body: 'The grill is down to temp; turning it off. The fan will run until the firepot cools.',
        },
      };
    }
    return { phase: 'cooling', action: null, notice: null };
  }
  if (phase === 'finishing') {
    // Fully off (module off AND cool-down fan stopped) — done. The recorder
    // announces the power-off, so we stay quiet here.
    if (!inp.moduleIsOn && !inp.fanState) return { phase: null, action: null, notice: null };
    return { phase: 'finishing', action: null, notice: null };
  }
  return { phase: null, action: null, notice: null };
}

// Cooling progress as a 0..1 fraction, from the temp cooling started at.
export function coolProgress(coolFrom: number, current: number): number {
  const span = coolFrom - SHUTDOWN.coolTarget;
  if (span <= 0) return 1;
  return Math.max(0, Math.min(1, (coolFrom - current) / span));
}

File: ios/PitBossKit/Sources/PitBossKit/Shutdown.swift
import Foundation

/// Graceful-shutdown decision logic — a direct port of `src/main/shutdown.ts`.
///
/// Why this matters: shutting a hot pellet grill straight off can let the fire
/// smoulder back up the auger toward the hopper — a burnback / hopper fire. On
/// this project's own grill that fire melted a bushing in the auger housing and
/// seized the auger motor, which had to be replaced. It is not a hypothetical.
///
/// The safe procedure is to bring the grill down to ~200°F first, *then* power
/// off so the controller's fan cool-down can fully extinguish the firepot.
///
/// Kept pure and dependency-free, exactly like the TypeScript original, so the
/// same assertions can be run against it (`pitboss-verify`). Thresholds are the
/// desktop's, which were validated on the grill on 2026-07-16 — they are not
/// re-guessed here.
public enum ShutdownPhase: Equatable, Sendable {
    case cooling
    case finishing
}

public struct ShutdownInput: Equatable, Sendable {
    public let moduleIsOn: Bool
    public let grillTemp: Int?
    public let grillSetTemp: Int?
    public let fanState: Bool

    public init(moduleIsOn: Bool, grillTemp: Int?, grillSetTemp: Int?, fanState: Bool) {
        self.moduleIsOn = moduleIsOn
        self.grillTemp = grillTemp
        self.grillSetTemp = grillSetTemp
        self.fanState = fanState
    }

    /// Reads the machine's inputs off a live grill state.
    public init(state: GrillState) {
        self.init(moduleIsOn: state.moduleIsOn == true,
                  grillTemp: state.grillTemp,
                  grillSetTemp: state.grillSetTemp,
                  fanState: state.fanState == true)
    }
}

public struct ShutdownStep: Equatable, Sendable {
    public enum Action: Equatable, Sendable { case cool, off }
    public struct Notice: Equatable, Sendable {
        public let title: String
        public let body: String
    }

    public let phase: ShutdownPhase?
    public let action: Action?
    public let notice: Notice?
}

public struct ShutdownConfig: Equatable, Sendable {
    /// Above this, ramp down before powering off.
    public let coolAbove: Int
    /// Ramp-down setpoint.
    public let coolTarget: Int
    /// Once at/below this, proceed to power off.
    public let coolDoneAt: Int
    /// How long before the cook is warned the cool-down hasn't finished.
    public let stallSeconds: TimeInterval

    public static let standard = ShutdownConfig(
        coolAbove: 250, coolTarget: 200, coolDoneAt: 210, stallSeconds: 30 * 60)

    public init(coolAbove: Int, coolTarget: Int, coolDoneAt: Int, stallSeconds: TimeInterval) {
        self.coolAbove = coolAbove
        self.coolTarget = coolTarget
        self.coolDoneAt = coolDoneAt
        self.stallSeconds = stallSeconds
    }
}

public enum Shutdown {
    /// The first step when the cook asks to shut down.
    public static func begin(_ input: ShutdownInput,
                             config: ShutdownConfig = .standard) -> ShutdownStep {
        if input.moduleIsOn, let temp = input.grillTemp, temp > config.coolAbove {
            return ShutdownStep(
                phase: .cooling,
                action: .cool,
                notice: .init(
                    title: "Cooling down before shutdown",
                    body: "Bringing the grill from \(temp)° to \(config.coolTarget)° first — this prevents a hopper flare-up."))
        }
        return ShutdownStep(
            phase: .finishing,
            action: .off,
            notice: .init(
                title: "Shutting down",
                body: "Turning the grill off; the fan will run until the firepot cools."))
    }

    /// Advances the machine on each fresh grill state.
    public static func advance(_ phase: ShutdownPhase?,
                               _ input: ShutdownInput,
                               config: ShutdownConfig = .standard) -> ShutdownStep {
        switch phase {
        case .cooling:
            if let temp = input.grillTemp, temp <= config.coolDoneAt {
                return ShutdownStep(
                    phase: .finishing,
                    action: .off,
                    notice: .init(
                        title: "Cooled — shutting down",
                        body: "The grill is down to temp; turning it off. The fan will run until the firepot cools."))
            }
            return ShutdownStep(phase: .cooling, action: nil, notice: nil)

        case .finishing:
            // Fully off (module off AND cool-down fan stopped) — done.
            if !input.moduleIsOn && !input.fanState {
                return ShutdownStep(phase: nil, action: nil, notice: nil)
            }
            return ShutdownStep(phase: .finishing, action: nil, notice: nil)

        case nil:
            return ShutdownStep(phase: nil, action: nil, notice: nil)
        }
    }

    /// Cooling progress as a 0…1 fraction, from the temp cooling started at.
    public static func coolProgress(from coolFrom: Int,
                                    current: Int,
                                    config: ShutdownConfig = .standard) -> Double {
        let span = Double(coolFrom - config.coolTarget)
        guard span > 0 else { return 1 }
        return max(0, min(1, Double(coolFrom - current) / span))
    }
}

/// What the cook asked for.
public enum ShutdownMode: Equatable, Sendable {
    /// Cool down first if hot, then power off.
    case auto
    /// Skip the cool-down — an explicit second press.
    case now
    /// Abort the shutdown and keep cooking.
    case cancel
}
